### DISCUSSION TITLE
Anyone have create chat dashboard with this ?

### DISCUSSION THREAD

**Discussion by bobykurniawan11** [devlikeapro/waha#710] [Q&A]
Hey everyone,

I’m working on a chat dashboard and could really use some advice. Specifically, I’m trying to figure out how to structure the database tables for stuff like:

Incoming messages – How do you store messages so it’s easy to load them in real-time?
New chats – What’s a good setup for tracking new conversations or chat sessions?
Extras – Things like read/unread status

If you’ve built something similar, I’d love to hear how you approached it. Any tips, examples, or lessons learned would be awesome!

Thanks in advance!

[![patron:PLUS](https://img.shields.io/badge/patron-PLUS-a0e6ba)](https://waha.devlike.pro/docs/how-to/plus-version/#tiers)

**Comment by devlikepro** [comment]
Hi! In 2024.12.6 we've added [**Chat UI example on Dashboard**](https://waha.devlike.pro/docs/how-to/dashboard/#chat-ui) so you can see how it can be done (it's just example, not full feature supported) 
![image](https://github.com/user-attachments/assets/e1debb6f-e267-45e3-8fbe-a94611a069c0)

![image](https://github.com/user-attachments/assets/fedf6bae-82a4-4da8-831a-612efd71b96b)

WAHA provides a new simple "chat overview" API basically for that purpose - so you can build alternative UI clients faster, without additions requests (we've been asked this feature many times) 
https://waha.devlike.pro/docs/how-to/chats/#get-chats-overview

----

Overall, you can just use WAHA HTTP API to show Chat UI, the database structure is well defined in WAHA,  status, etc you can consume from API responses.
Here's few things you need to take care if you want to reimplement it on your side

> **Incoming messages** – How do you store messages so it’s easy to load them in real-time?

So you'll need to have indexes over
1. `chat.id` 
2. `timestamp` 
3. `chat.id + timestamp`
4. `message.id`

We use key-value approach even in sql databases - extract necessary fields in separate columns, create index and save in `data` json representation of the entity (chat/message/group/etc)

-----

>  New chats – What’s a **good setup for tracking new conversations** or chat sessions?

[**Websockets**](https://waha.devlike.pro/docs/how-to/webhooks/#websockets) is must have for good UX.
:warning: remember that websocket waha exposes works with **master api key**, so you likely will need to setup additional server to authenticate your clients properly on your side.

We'll need to work in some temp token direction for that purpose, like https://github.com/devlikeapro/waha/issues/643 
With that setup, **on your server** you can ask WAHA to get you JWT token for websocket, so UI client can connect to WAHA directly, without additional middleware. Hit +1 in the comment so we know how many people are interested in that and get notification when it's ready!

----

> Extras – Things like read/unread status

Listen to `message.ack` events and update it properly on UI side. We don't recommend requesting messages when you get any new `message.any` or `message.ack` event, but it's fine for apps with low message rate





[![patron:PRO](https://img.shields.io/badge/patron-PRO-188a42)](https://waha.devlike.pro/docs/how-to/plus-version/#tiers)

**Comment by bobykurniawan11** [comment]
Thank you so much for your detailed answer!

[![patron:PLUS](https://img.shields.io/badge/patron-PLUS-a0e6ba)](https://waha.devlike.pro/docs/how-to/plus-version/#tiers)
